src/light_data.rs

```diff
@@ -0,0 +1,307 @@
+use std::fmt::Display;
+
+#[cfg(not(target_arch = "wasm32"))]
+use proptest::prelude::*;
+#[cfg(not(target_arch = "wasm32"))]
+use rand::{rngs::StdRng, SeedableRng};
+
+use nom::bytes::complete::take;
+use nom::multi::count;
+use nom::Finish;
+use nom::IResult;
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub enum LightData {
+    Atom(Vec<u8>),
+    Cell(Vec<LightData>),
+}
+
+impl Display for LightData {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Atom(xs) => {
+                write!(f, "[")?;
+                let mut iter = xs.iter().peekable();
+                while let Some(x) = iter.next() {
+                    write!(f, "{:02x?}", x)?;
+                    if let Some(_) = iter.peek() {
+                        write!(f, ", ")?;
+                    } else {
+                        write!(f, "]")?;
+                    }
+                }
+                Ok(())
+            }
+            Self::Cell(xs) => {
+                write!(f, "(")?;
+                let mut iter = xs.iter().peekable();
+                while let Some(x) = iter.next() {
+                    write!(f, "{}", x)?;
+                    if let Some(_) = iter.peek() {
+                        write!(f, ", ")?;
+                    } else {
+                        write!(f, ")")?;
+                    }
+                }
+                Ok(())
+            }
+        }
+    }
+}
+
+impl Arbitrary for LightData {
+    type Parameters = ();
+    type Strategy = BoxedStrategy<Self>;
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
```

Comment:
This won’t produce any top-level Atom, but I’ll just chalk that up to the WIP state of the PR.